### DISCUSSION TITLE
XlmRoBertaSentenceEmbeddings returns huge amount of embeddings instead of set dimensions

### DISCUSSION THREAD

**Discussion by kkwasnioch** [JohnSnowLabs/spark-nlp#14180] [Q&A]
I am trying to produce embeddings for whole documents in 3 languages: english, polish, finnish. Previously I have tried sentence-transformers/paraphrase-multilingual-mpnet-base-v2 from huggingface and it works fine, returns 768 dims. But when I load model and run it with sparknlp XlmRoBertaSentenceEmbeddings it produce f.e. 26k dims. Am I loading model wrong way? Or are thare any othe issues? Thanks!
https://github.com/JohnSnowLabs/spark-nlp/blob/master/examples/python/transformers/onnx/HuggingFace_ONNX_in_Spark_NLP_XlmRoBertaSentenceEmbeddings.ipynb -> here is sample code which i took knowladge
Code:

```
MODEL_NAME = "sentence-transformers/paraphrase-multilingual-mpnet-base-v2"
EXPORT_PATH = f"onnx_models/{MODEL_NAME}"
robert = XlmRoBertaSentenceEmbeddings.loadSavedModel(f"{EXPORT_PATH}", spark)\
    .setInputCols(["document"])\
    .setOutputCol("embeddings")\
    .setStorageRef('xlmroberta_embeddings_paraphrase_mpnet_base_v2') 

document_assembler = DocumentAssembler() \
    .setInputCol("text") \
    .setOutputCol("document")

embeddings_finisher = EmbeddingsFinisher() \
  .setInputCols('embeddings') \
  .setOutputCols('finnished_vectors') \
  .setOutputAsVector(False)

pipeline = Pipeline(stages=[document_assembler, robert, embeddings_finisher])

pipelineModel = pipeline.fit(sparkDF)
LightPipelinelightModel = LightPipeline(pipelineModel, parse_embeddings=True)

out = LightPipelinelightModel.transform(sparkDF).select('text', f.explode('finnished_vectors').alias('emb')).withColumn('size', f.size('emb'))
```

Output:
+--------------------+--------------------+-----+
|                             text|                           emb|   size|
+--------------------+--------------------+-----+
|Do kościoła jak "...     |[0.028680567, 0.2...|29952|
|Audi Q7 właśnie p...   |[-0.01756316, -0.... |28416|
|Białoruś. KGB wpr...   |[0.07118901, -0.0... |28416|
|"Są prawdziwym za...|[0.0972352, -0.04..|25344|
|Obsesja, za którą...    |[0.07850968, 0.15..|32256|
|Ogromny sukces Po...|[-0.034644652, 0..|22272|
|Rolnicy \"zajęli\...       |[-0.06938014, 0.0.. |29952|
|Szokujące wyznani... |[0.08084734, 0.18...|30720|
|Pogoda zaskoczy w...|[-0.086600736, 0....|34560|
|Kiedyś kary fizyc...     |[0.059363756, 0.0..|28416|
+--------------------+--------------------+-----+

**Comment by maziyarpanahi** [comment]
Hi @kkwasnioch 
This seems to be a bug that was introduce in other ONNX-based annotators dealing with sentence embeddings. We fix them for MPNet and E5, but we missed this one and probably other `XXXSentenceEmbeddings` in Spark NLP.

Will try to fix this in the next release

**Reply by kkwasnioch** [reply]
@maziyarpanahi thanks for the answer! Are there any expected dates for the next release?

**Reply by maziyarpanahi** [reply]
If we can fix this before the next one I would say it is targeted for this week. In the meantime, you can use MPNet that has similar models for multilingual feature and once we fix the XLM-RoBERTa for sentence embeddings you can simple switch the embeddings stage: https://github.com/JohnSnowLabs/spark-nlp/blob/master/examples/python/transformers/onnx/HuggingFace_ONNX_in_Spark_NLP_MPNet.ipynb

PS: I also just realized we never uploaded XLM-RoBERTa notebooks, you could have used the `XlmRoBertaEmbeddings` which is a token embeddings with `SentenceEmbeddings` that averages all the tokens.

**Reply by kkwasnioch** [reply]
Oh I didn't realize that MPNET all-mpnet-base-v2 is multilingual, thought it only works for english, I will defenilty try it out! I used XlmRoBertaEmbeddings with SentenceEmbeddings but somehow it returns other embedding values than the regular one from huggingface. I measured it with this piece:

```
cosine_similarity(np.array(vec_df_regular_python['embeddings_pl'].iloc[1]).reshape(1,-1), 
                  np.array(vec_df_sparknlp['embeddings_pl'].iloc[1]).reshape(1,-1))[0][0]
```
output: ~0.6505

**Reply by maziyarpanahi** [reply]
That's true, `SentenceEmbeddings` is a very simple average rather than a average_pool like MPNet or E5 or INSTRUCT. It was implemented initially for Glove and Word2Vec, but it can be used with other context-aware LLMs like XLM-RoBERTa. (although, the accuracy is a bit lower with a simple average).

But MPNet uses the original average pool.

**Reply by kkwasnioch** [reply]
It make sense. Thanks! Are there any dedicated models for multilingual tasks from MPNet or I can simply just use the one proposed in colab notebook?

**Reply by maziyarpanahi** [reply]
not that I am aware of, most people use XLM-RoBERTa arch to offer multilingual model. Sentence Transformers did the same with `paraphrase-multilingual-mpnet-base-v2`. But we don have these models for MPNet if anyone them were useful: https://sparknlp.org/models?annotator=MPNetEmbeddings

**Reply by kkwasnioch** [reply]
paraphrase-multilingual-mpnet-base-v2 is the one I would like to use in future, it seems to returns best results. Should it work with fixed XlmRoBertaSentenceEmbeddings module?

**Reply by maziyarpanahi** [reply]
100%! once we fix it will work without any issue

**Reply by kkwasnioch** [reply]
@maziyarpanahi Greetings! Have you already implemented the fixes to issue? :)
